src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,202 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> currentNode;
+    private Node<T> previousNode;
+    private Node<T> nextNode;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> newElement = new Node<T>(value);
+            head = newElement;
+            tail = newElement;
+        } else {
+            Node<T> newElement = new Node<T>(value);
+            newElement.prev = tail;
+            Node<T> previousNode = tail;
+            previousNode.next = newElement;
+            tail = newElement;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("The index passed must be greater or equal to 0 "
+                    + "and less or equal to" + size);
+        }
+        if (size - index == 0 || size == 0) {
+            add(value);
+            return;
+        }
+        iterateCurrentNode(index);
```

Comment:
The method 'iterateCurrentNode' iterates from the head to the specified index. Consider optimizing this by iterating only the first or second half of the list, depending on the index value.